examples/benchmark/datagen.py

```diff
@@ -0,0 +1,205 @@
+#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#  // SPDX-License-Identifier: BSD
+import abc
+import os
+import tarfile
+import tempfile
+from pathlib import Path
+from typing import Generator
+
+import boto3
+import click
+import numpy as np
+import prefixed
+import prefixed as pr
+import yaml
+from PIL import Image
+from joblib import parallel_config, Parallel, delayed, cpu_count
+from tqdm import tqdm
+
+
+class DataGenerator(abc.ABC):
+    @abc.abstractmethod
+    def generate(self, index: int, destdir: Path) -> bytes:
+        pass
+
+
+class ImageGenerator(DataGenerator):
+    def __init__(self, width: int, height: int, img_format: str = "JPEG"):
+        self.width = width
+        self.height = height
+        self.img_format = img_format
+
+    def generate(self, index: int, destdir: Path):
+        img = Image.fromarray(
+            np.random.randint(
+                0, high=256, size=(self.height, self.width, 3), dtype=np.uint8
+            )
+        )
+        img_file: Path = destdir / f"image_{index}.{self.img_format}"
```

Comment:
Why are we saving these files to disk rather than using `BytesIO`?

Reply:
The original version was written to synthesize files on disk and rely on `aws s3 sync` which would handle the upload externally. 

Since we're handling uploads as well now, we should ideally be implementing it as a a pipeline like what you're suggesting(especially if datasets are to get really large).

I'm thinking it should be reasonable to address this in a future iteration and take this down as a planned enhancement. Let me know what you think.